### ISSUE TITLE
back compat serialization

### ISSUE THREAD

**Issue by soumith** [torch/nn#503]
add tests for back-compat serialization for all nn modules.

**Comment by andreaskoepf** [comment]
Could we introduce a new (optional) method that gets called after deserialization of an object to check and potentially correct the object... For module authors it might not be obvious that there are ways to (re-)create the objects which do not run the ctor.

**Comment by soumith** [comment]
We already have custom read / write that can be specified. I did it this
way in a hurry. More proper fix and unit tests will come over the weekend.

**Comment by andresy** [comment]
If it can be of any help: do you know that :read(file, [version]) takes an optional argument version number (which reflects __version in the metatable), such that you can handle new serialisation variants?

**Comment by andreaskoepf** [comment]
@soumith & @andresy thnx for mentioning read which I did not know.

I created a [mini demo gist](https://gist.github.com/andreaskoepf/8a2df7daea695d98d062#file-torch_obj_versioning-lua) which shows a `restore` function that is called from a custom read impl after simply assigning all deserialized-members without changes.

**Comment by andreaskoepf** [comment]
Maybe it would be possible to have a method similar to `__factory` but that is called during normal class construction and during deserialization. Or maybe we could add a flag in the class-metatable that would trigger execution of the ctor before deserialization. This would at least cover the very common case of additive changes: e.g. replace the need for `backCompatibility()` in  [LookupTable.lua](https://github.com/torch/nn/blob/master/LookupTable.lua).
In general I think it would be better to 'upgrade' an obj once after deserialization than having backwardCompatibility() calls everywhere.

**Comment by andresy** [comment]
you could have a subinit() function called from your custom __factory() as well as the constructor.

**Comment by andreaskoepf** [comment]
@andresy Yes, I was thinking of same kind of 'default ctor' to reduce code duplication in ctor and the custom-deserialization handler - but I forgot that many initializations depend on parameters or state: E.g. a common case would be to add a new tensor field - but it could not be created in a default ctor because its type has to be the same as the other tensors after deserialization (e.g. if the module was  `:cuda()`-converted before serialization): So after sleeping over it I am no longer convinced a default-ctor `subinit()` would be of much help. The `restore()` method solution shown in my gist might be a candidate: e.g. we could add the custom `read` method as `nn.Module:read()` which would call an optional `restore()` method if it was present...

**Comment by andreaskoepf** [comment]
All classes derived from `nn.Module' have at least empty `gradInput` and `output` tensors that hold tensor type information and could be used to initialize new tensors via `self.newFieldXyz = self.newFieldXyz or self.output:new()`. In order for this to work the upcoming `clearStates()` method (mentioned by @soumith on gitter) must not completely remove the fields but 'clear' them, e.g. by calling `self.gradInput:set()` (set method without arg).
